rpc: add getorphantxs

tdb3 opened the pull request:
This PR adds a new hidden rpc, `getorphantxs`, that provides the caller with a list of orphan transactions.  This rpc may be helpful when checking orphan behavior/scenarios (e.g. in tests like `p2p_orphan_handling`) or providing additional data for statistics/visualization.

```
getorphantxs ( verbosity )

Shows transactions in the tx orphanage.

EXPERIMENTAL warning: this call may be changed in future releases.

Arguments:
1. verbosity    (numeric, optional, default=0) 0 for an array of txids (may contain duplicates), 1 for an array of objects with tx details, and 2 for details from (1) and tx hex

Result (for verbose = 0):
[           (json array)
  "hex",    (string) The transaction hash in hex
  ...
]

Result (for verbose = 1):
[                          (json array)
  {                        (json object)
    "txid" : "hex",        (string) The transaction hash in hex
    "wtxid" : "hex",       (string) The transaction witness hash in hex
    "bytes" : n,           (numeric) The serialized transaction size in bytes
    "vsize" : n,           (numeric) The virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted.
    "weight" : n,          (numeric) The transaction weight as defined in BIP 141.
    "expiration" : xxx,    (numeric) The orphan expiration time expressed in UNIX epoch time
    "from" : [             (json array)
      n,                   (numeric) Peer ID
      ...
    ]
  },
  ...
]

Result (for verbose = 2):
[                          (json array)
  {                        (json object)
    "txid" : "hex",        (string) The transaction hash in hex
    "wtxid" : "hex",       (string) The transaction witness hash in hex
    "bytes" : n,           (numeric) The serialized transaction size in bytes
    "vsize" : n,           (numeric) The virtual transaction size as defined in BIP 141. This is different from actual serialized size for witness transactions as witness data is discounted.
    "weight" : n,          (numeric) The transaction weight as defined in BIP 141.
    "expiration" : xxx,    (numeric) The orphan expiration time expressed in UNIX epoch time
    "from" : [             (json array)
      n,                   (numeric) Peer ID
      ...
    ],
    "hex" : "hex"          (string) The serialized, hex-encoded transaction data
  },
  ...
]

Examples:
> bitcoin-cli getorphantxs 2
> curl --user myusername --data-binary '{"jsonrpc": "2.0", "id": "curltest", "method": "getorphantxs", "params": [2]}' -H 'content-type: application/json' http://127.0.0.1:8332/
```
```
$ build/src/bitcoin-cli getorphantxs 2
[
  {
    "txid": "50128aac5deab548228d74d846675ad4def91cd92453d81a2daa778df12a63f2",
    "wtxid": "bb61659336f59fcf23acb47c05dc4bbea63ab533a98c412f3a12cb813308d52c",
    "bytes": 133,
    "vsize": 104,
    "weight": 415,
    "expiration": 1725663854,
    "from": [
      1
    ],
    "hex": "020000000001010b992959eaa2018bbf31a4a3f9aa30896a8144dbd5cfaf263bf07c0845a3a6620000000000000000000140fe042a010000002251202913b252fe537830f843bfdc5fa7d20ba48639a87c86ff837b92d083c55ad7c102015121c0000000000000000000000000000000000000000000000000000000000000000100000000"
  },
  {
    "txid": "330bb7f701604a40ade20aa129e9a3eb8a7bf024e599084ca1026d3222b9f8a1",
    "wtxid": "b7651f7d4c1a40c4d01f6a1e43a121967091fa0f56bb460146c1c5c068e824f6",
    "bytes": 133,
    "vsize": 104,
    "weight": 415,
    "expiration": 1725663854,
    "from": [
      2
    ],
    "hex": "020000000001013600adfe41e0ebd2454838963d270916d2b47239c9eebb93a992b720d3589a080000000000000000000140fe042a010000002251202913b252fe537830f843bfdc5fa7d20ba48639a87c86ff837b92d083c55ad7c102015121c0000000000000000000000000000000000000000000000000000000000000000100000000"
  }
]
```

andrewtoth commented:
Can you expand on the motivation behind exposing this please? Why would someone want to use such an RPC?

tdb3 commented:
Thought more about this and other than increased visibility and monitoring of the orphanage over time (e.g. for statistical purposes), I'm not seeing a ton of compelling reasons to add.  Will close for now.

glozow reviewed (commented):
Concept ACK

I once talked with @0xB10C about adding this, but never got around to it. I think it would be helpful for visualizing / trying out different orphan protection methods.

tdb3 commented on src/net_processing.cpp:
```diff
@@ -1920,6 +1921,18 @@ bool PeerManagerImpl::GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) c
     return true;
 }
 
+std::vector<CTransaction> PeerManagerImpl::GetOrphans()
```
Right. Initially, the thought was to break clean from PeerManager/TxOrphanage, but CTransactionRef is a shared_ptr so could access the CTransaction even if it was dropped in TxOrphanage.

tdb3 commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +813,46 @@ static RPCHelpMan savemempool()
+static RPCHelpMan getorphantxs()
+{
+    return RPCHelpMan{"getorphantxs",
+        "\nShows transactions from the tx orphanage.\n",
```
Thank you.  Both of those suggestions are reasonable to me.  Will update soon.  Do you have a preference?

tdb3 replied:
Updated

instagibbs commented:
this PR might be a good time to enhance `test/functional/p2p_orphan_handling.py`, reducing our reliance on log messages at least

tdb3 commented:
> this PR might be a good time to enhance `test/functional/p2p_orphan_handling.py`, reducing our reliance on log messages at least

Great idea. For now, a child branch was started (https://github.com/tdb3/bitcoin/tree/enhance_p2p_orphan_handling) that adds a commit (https://github.com/tdb3/bitcoin/commit/344ab30487235a2096d878894dc6f4e2d0d1ef65) to update `p2p_orphan_handling`.

Since additional enhancements could be made to that branch (e.g. to check the orphanage when the debug log is silent), maybe we keep this PR scoped to the new RPC, then a followup PR uses the `enhance_p2p_orphan_handling` branch?  This could simplify review of both PRs.

Alternatively, we could add just the commit handling `assert_debug_log` reduction here, and the followup would further enhance `p2p_orphan_handling`.

tdb3 commented:
Updated to include an experimental warning and make the RPC hidden.
Also added a convenience function for checking the orphanage in tests.

0xB10C commented:
> Could also look into adding more transaction details (e.g. orphan expiration time, and potentially other tx information similar to what's provided in getrawmempool or getrawtransaction).

> I think it would be helpful for visualizing / trying out different orphan protection methods.

Concept ACK

I agree that this would be interesting. The the orphan pool is mostly a black box to me (and probably others too). To discuss improvements/changes, having a better visual and mental model would be helpful. Maybe even similar to `getrawaddrman` and https://addrman.observer.

tdb3 commented:
> Concept ACK
> 
> I agree that this would be interesting. The the orphan pool is mostly a black box to me (and probably others too). To discuss improvements/changes, having a better visual and mental model would be helpful. Maybe even similar to `getrawaddrman` and https://addrman.observer.

Thanks.  That web app is gorgeous!  Any particular tx details you'd like to see added (e.g. orphan expiration, etc.)?

instagibbs commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +813,47 @@ static RPCHelpMan savemempool()
+    UniValue ret(UniValue::VARR);
+    for (auto const& orphan : orphanage) {
+        UniValue o(UniValue::VOBJ);
+        o.pushKV("txid", orphan->GetHash().ToString());
```
I feel like the ability to fetch the whole transaction (set) optionally is useful, otherwise we just have a couple hashes of the transactions, no ability to estimate how many vbytes the orphanage is using, do any analysis on the data given?

Maybe have an optional argument via txid/wtxid that fetches the full serialized transaction in hex?

tdb3 replied:
Agreed, will refinea

instagibbs replied:
Follow-up PRs can probably add more interesting fields as well, such as entry time, announcer peerid, etc.

I'll try not to scope creep this PR too bad

glozow replied:
> ability to estimate how many vbytes the orphanage is using

A `getorphanageinfo` can be added for overall stats, perhaps after #28031 which adds size tracking.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +813,39 @@ static RPCHelpMan savemempool()
+            RPCResult{"orphans",
+                RPCResult::Type::ARR, "", "",
+                {
+                    {RPCResult::Type::STR_HEX, "", "The transaction id"},
```
For originator, I was thinking `fromPeer` (the NodeId). Currently each orphan tracks 1 sender, but hoping to have multiple in the future (#28031 if you're curious).

For this PR, I think it would suffice to take an approach similar to `getrawmempool` but with an integer verbosity.
verbose=0: list of txids (it's ok if there are duplicates, it should be rare anyway)
verbose=1: list of objects with txid, wtxid, size, senders (a list of NodeIds which as of now has length 1), expiration

I think a further verbose=2 could contain the transaction hex as well. There are more orphanage changes coming, which could add more fields. However I don't want to hold things up with bikeshedding, so feel free to decide what you think the scope of this PR should be.

tdb3 replied:
Thanks.  Updated with three verbosity levels (0, 1, and 2).

tdb3 committed:
refactor: add OrphanTxBase for external use

Enables external entities to obtain orphan
information

tdb3 commented:
The push to f6dee50f85b2668e96e9dffd65e06d8f3d3a4da2 adds verbosity levels and more detailed information about the orphans.

tdb3 commented:
Not sure if the previous releases CI failure (timeout) is unrelated (https://cirrus-ci.com/task/6175438823751680).  Nothing sticks out immediately to me, except for possibly the following (but maybe this is just a symptom?):
```
+ podman container rm --force --all
time="2024-09-07T09:55:54Z" level=warning msg="StopSignal SIGTERM failed to stop container ci_native_previous_releases in 10 seconds, resorting to SIGKILL"
ef7ef7197494aa565138bc0560b4937dd11a79ade74357aa89bbf111955959e0
```

Also, the 32-bit CentOS CI barely passed (https://cirrus-ci.com/task/5310638300332032).  I clocked compiling finishing in 85 minutes, with unit tests completing about 15 minutes later and functionals passing just shy of the 120-minute mark.

Thoughts @maflcko?

brunoerg commented:
Concept ACK

glozow commented on src/txorphanage.h:
```diff
@@ -81,6 +81,13 @@ class TxOrphanage {
         NodeSeconds nTimeExpire;
     };
 
+    /**
+     * Provides orphans
+     *
+     * @returns Orphans with metadata
+     */
```
Maybe self-explanatory :sweat_smile:

glozow commented on src/net_processing.cpp:
```diff
@@ -519,6 +519,7 @@ class PeerManagerImpl final : public PeerManager
     std::optional<std::string> FetchBlock(NodeId peer_id, const CBlockIndex& block_index) override
         EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
     bool GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const override EXCLUSIVE_LOCKS_REQUIRED(!m_peer_mutex);
+    std::vector<TxOrphanage::OrphanTxBase> GetOrphans() override EXCLUSIVE_LOCKS_REQUIRED(!m_tx_download_mutex);
```
Perhaps `GetOrphanTransactions`, since this is on the peerman and orphan blocks can exist too

tdb3 replied:
Agreed. Updated.

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+    return {
+        RPCResult{RPCResult::Type::STR_HEX, "txid", "The transaction hash in hex"},
+        RPCResult{RPCResult::Type::STR_HEX, "wtxid", "The transaction witness hash in hex"},
+        RPCResult{RPCResult::Type::NUM, "size", "The serialized transaction size"},
```
maybe "bytes" instead? to make it hard to confuse with vsize

tdb3 replied:
"bytes" seems clearer to me as well. Updated.

Originally "size" was chosen because it aligns with `getrawtransaction`.
If there's friction with using "bytes", I can change it back to "size".

> Argument and field naming: please consider whether there is already a naming style or spelling convention in the API for the type of object in question (blockhash, for example), and if so, try to use that

https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#rpc-interface-guidelines

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+        RPCResult{RPCResult::Type::STR_HEX, "wtxid", "The transaction witness hash in hex"},
+        RPCResult{RPCResult::Type::NUM, "size", "The serialized transaction size"},
+        RPCResult{RPCResult::Type::NUM, "vsize", "The virtual transaction size (differs from size for witness transactions)"},
+        RPCResult{RPCResult::Type::NUM, "weight", "The transaction weight"},
```
mention BIP 141?

tdb3 replied:
Thanks. Increases clarity. Updated.

tdb3 replied:
Thanks. Increases clarity. Updated.

glozow commented on test/functional/rpc_getorphantxs.py:
```diff
@@ -0,0 +1,122 @@
+        assert orphanage[0]["from"][0] != orphanage[1]["from"][0]
+
+        self.log.info("Unorphan child 2")
+        peer_2.send_and_ping(msg_tx(tx_parent_2["tx"]))
```
Maybe should have a method for testing that a tx isn't in orphanage so we can test that here

tdb3 replied:
Seemed reasonable to add a check `assert not node.tx_in_orphanage(tx_child_2["tx"])` rather than a new method, but happy to be convinced otherwise.

glozow commented on test/functional/rpc_getorphantxs.py:
```diff
@@ -0,0 +1,122 @@
+
+        self.log.info("Checking orphan details (verbosity 1)")
+        orphanage = node.getorphantxs(1)
+        orphan_1 = orphanage[0]
```
I was briefly confused about ordering stability, and then realized this has length 1

tdb3 replied:
Thanks, yeah it's a bit too implicit so I added `assert_equal(len(node.getorphantxs()), 1)` to make it clearer and more explicit.

tdb3 committed:
net: add GetOrphanTransactions() to PeerManager

Updates PeerManager (and Impl) to provide
orphans with metadata

tdb3 committed:
refactor: move verbosity parsing to rpc/util

Provides a common way for rpcs to obtain
verbosity from an rpc parameter

tdb3 committed:
rpc: add getorphantxs

Adds an rpc to obtain data about the
transactions currently in the orphanage.
Hidden and marked as experimental

tdb3 committed:
test: add tx_in_orphanage()

Allows tests to check if a transaction
is contained within the orphanage

tdb3 committed:
test: add getorphantxs tests

Adds functional tests for getorphantxs

tdb3 commented:
Thanks for the review @glozow!

hodlinator commented on test/functional/rpc_getorphantxs.py:
```diff
@@ -0,0 +1,124 @@
+        self.log.info("Check that the orphanage is empty")
+        assert_equal(len(node.getorphantxs()), 0)
+
+        self.log.info("Confirm the transactions (clears mempool)")
+        self.generate(node, 1)
```
nit: Might be nice to check the mempool size is zero as a post-condition, even though it should be covered by other tests.

tdb3 replied:
Added check

hodlinator reviewed (commented):
Concept ACK

Feels like *p2p_orphan_handling.py*/`test_same_txid_orphan()` might benefit from some augmented checking using this new `getorphantxs`. It could verify that same-txids-differing-wtxid orphans are handled in the expected way, with the bogus one being removed after the one with the valid witness gets adopted.

0xB10C commented:
Played around with the `getorphantxs 1` output a bit regarding visualization. You can now plug your own `getorphantxs.json` (verbosity >=1) into this demo tool https://observablehq.com/d/fb8142e7eb7f98f3. Produces something like this based on the size and from-peers (color) of the orphans.

![image](https://github.com/user-attachments/assets/fb1d4d9e-8e77-4e53-89fe-ea42fc5b1847)

tdb3 commented:
> Concept ACK
> 
> Feels like _p2p_orphan_handling.py_/`test_same_txid_orphan()` might benefit from some augmented checking using this new `getorphantxs`. It could verify that same-txids-differing-wtxid orphans are handled in the expected way, with the bogus one being removed after the one with the valid witness gets adopted.

Great idea. Agreed. A branch was started (https://github.com/bitcoin/bitcoin/pull/30793#issuecomment-2327703573) to capture changes to `p2p_orphan_handling.py`. Updates to `p2p_orphan_handling.py` are planned for a follow-up PR to keep this one scoped and easier to review.

tdb3 commented on test/functional/test_framework/test_node.py:
```diff
@@ -841,6 +844,10 @@ def bumpmocktime(self, seconds):
     def wait_until(self, test_function, timeout=60):
         return wait_until_helper_internal(test_function, timeout=timeout, timeout_factor=self.timeout_factor)
 
+    def tx_in_orphanage(self, tx: CTransaction) -> bool:
+        """Returns true if the transaction is in the orphanage."""
+        found = [o for o in self.getorphantxs(verbosity=1) if o["txid"] == tx.rehash() and o["wtxid"] == tx.getwtxid()]
+        return len(found) == 1
```
The original thought/rationale was that the function pertains to the orphanage, which is part of the node (although a `node` argument to the function could enable obtaining the same information).
No file stands out as a perfect fit.  Do you feel `test_framework/mempool_util.py` would be reasonable? It seems appropriate since the orphanage and mempool both deal with unconfirmed transactions and are somewhat related.

Let me know and I can update either way.

hodlinator replied:
Ah, yeah, `mempool_util.py` looks like a better alternative than `blocktools.py`.

`test_node.py` feels more like the high-level P2P stuff to me at least + lower level dynamic RPC invocation.

tdb3 replied:
Moved to `mempool_util.py`

tdb3 reviewed (commented):
Thanks for the review @hodlinator!  Left some updates.

tdb3 commented:
> Played around with the `getorphantxs 1` output a bit regarding visualization. You can now plug your own `getorphantxs.json` (verbosity >=1) into this demo tool https://observablehq.com/d/a481f4ced64b7975. Produces something like this based on the size and from-peers (color) of the orphans.

This looks great!

itornaza commented on src/txorphanage.h:
```diff
@@ -72,11 +72,19 @@ class TxOrphanage {
         NodeSeconds nTimeExpire;
+    };
+
+    std::vector<OrphanTxBase> GetOrphanTransactions();
```
Do you think defining this as a const function `std::vector<OrphanTxBase> GetOrphanTransactions() const;` makes sense? I do not think it changes any members of the `TxOrphanage`.

tdb3 replied:
Thanks! Will include this in next push.

tdb3 replied:
Updated

itornaza commented on src/txorphanage.cpp:
```diff
@@ -277,3 +277,13 @@ std::vector<std::pair<CTransactionRef, NodeId>> TxOrphanage::GetChildrenFromDiff
     return children_found;
 }
+
+std::vector<TxOrphanage::OrphanTxBase> TxOrphanage::GetOrphanTransactions()
```
My comment in the function declaration at the `txorphanage.h:84` applies here as well.

tdb3 replied:
Updated

itornaza reviewed (approved):
Concept ACK.

The changes look good to me and the added functionality looks even greater! Built and run the extended test harness and everything checks out. Left more like a question rather than a nit.

hodlinator commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+            HelpExampleCli("getorphantxs", "2")
+            + HelpExampleRpc("getorphantxs", "2")
+        },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
```
Exactly like that, thanks!

tdb3 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -338,15 +338,7 @@ static RPCHelpMan getrawtransaction()
-            verbosity = request.params[1].getInt<int>();
-        }
-    }
+    int verbosity{ParseVerbosity(request.params[1])};
```
Updated

tdb3 commented:
Updated to e6853592361341c27103ed74b25470ac1e098d6d to address comments from @hodlinator and @itornaza

tdb3 commented:
Pushed to b6368fc285bf00b3033061dcd4e29298b227c6df (rebased on top of fc642c33ef28829eda0119a0fe39fd9bc4b84051)

<details>
<summary>Saw the following CI error for macOS 14 with e6853592361341c27103ed74b25470ac1e098d6d</summary>

This seems to be related to PR #30684

```
node0 2024-10-02T00:15:45.783638Z [init] [src/noui.cpp:57] [noui_InitMessage] init message: Verifying wallet(s)… 
Error:  node0 2024-10-02T00:15:45.783653Z [init] [src/noui.cpp:31] [noui_ThreadSafeMessageBox] [error] Invalid value detected for '-wallet' or '-nowallet'. '-wallet' requires a string value, while '-nowallet' accepts only '1' to disable all wallets 
 node0 2024-10-02T00:15:45.783737Z [init] [src/httpserver.cpp:510] [InterruptHTTPServer] [http] Interrupting HTTP server 
 node0 2024-10-02T00:15:45.783748Z [init] [src/httprpc.cpp:379] [InterruptHTTPRPC] [rpc] Interrupting HTTP RPC server 
 node0 2024-10-02T00:15:45.783756Z [init] [src/rpc/server.cpp:291] [operator()] [rpc] Interrupting RPC 
 node0 2024-10-02T00:15:45.783770Z [init] [src/init.cpp:277] [Shutdown] Shutdown: In progress... 
 node0 2024-10-02T00:15:45.783782Z [shutoff] [src/httprpc.cpp:384] [StopHTTPRPC] [rpc] Stopping HTTP RPC server 
 node0 2024-10-02T00:15:45.783794Z [shutoff] [src/httpserver.cpp:772] [UnregisterHTTPHandler] [http] Unregistering HTTP handler for / (exactmatch 1) 
 node0 2024-10-02T00:15:45.783804Z [shutoff] [src/httpserver.cpp:772] [UnregisterHTTPHandler] [http] Unregistering HTTP handler for /wallet/ (exactmatch 0) 
 node0 2024-10-02T00:15:45.783810Z [shutoff] [src/rpc/server.cpp:303] [operator()] [rpc] Stopping RPC 
 node0 2024-10-02T00:15:45.783892Z [shutoff] [src/rpc/server.cpp:309] [operator()] [rpc] RPC stopped. 
 node0 2024-10-02T00:15:45.783904Z [shutoff] [src/httpserver.cpp:522] [StopHTTPServer] [http] Stopping HTTP server 
 node0 2024-10-02T00:15:45.783911Z [shutoff] [src/httpserver.cpp:524] [StopHTTPServer] [http] Waiting for HTTP worker threads to exit 
 node0 2024-10-02T00:15:45.784048Z [http] [src/httpserver.cpp:354] [ThreadHTTP] [http] Entering http event loop 
 node0 2024-10-02T00:15:45.784129Z [http] [src/httpserver.cpp:357] [ThreadHTTP] [http] Exited http event loop 
 test  2024-10-02T00:55:45.674000Z TestFramework (ERROR): Assertion failed 
                                   Traceback (most recent call last):
                                     File "/Users/runner/work/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 665, in assert_start_raises_init_error
                                       ret = self.process.wait(timeout=self.rpc_timeout)
                                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 1264, in wait
                                       return self._wait(timeout=timeout)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/opt/homebrew/Cellar/python@3.12/3.12.6/Frameworks/Python.framework/Versions/3.12/lib/python3.12/subprocess.py", line 2045, in _wait
                                       raise TimeoutExpired(self.args, timeout)
                                   subprocess.TimeoutExpired: Command '['/Users/runner/work/bitcoin/bitcoin/ci/scratch/build-aarch64-apple-darwin23.6.0/src/bitcoind', '-datadir=/Users/runner/work/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20241002_001101/feature_settings_11/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-debugexclude=rand', '-uacomment=testnode0', '-logthreadnames', '-logsourcelocations', '-loglevel=trace', '-v2transport=0', '-settings=/Users/runner/work/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20241002_001101/feature_settings_11/node0/regtest/settings.json']' timed out after 2400 seconds
                                   During handling of the above exception, another exception occurred:
                                   Traceback (most recent call last):
                                     File "/Users/runner/work/bitcoin/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build-aarch64-apple-darwin23.6.0/test/functional/feature_settings.py", line 110, in run_test
                                       self.test_wallet_settings(settings)
                                     File "/Users/runner/work/bitcoin/bitcoin/ci/scratch/build-aarch64-apple-darwin23.6.0/test/functional/feature_settings.py", line 39, in test_wallet_settings
                                       node.assert_start_raises_init_error(expected_msg="Error: Invalid value detected for '-wallet' or '-nowallet'. '-wallet' requires a string value, while '-nowallet' accepts only '1' to disable all wallets",
                                     File "/Users/runner/work/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 695, in assert_start_raises_init_error
                                       self._raise_assertion_error(assert_msg)
                                     File "/Users/runner/work/bitcoin/bitcoin/test/functional/test_framework/test_node.py", line 197, in _raise_assertion_error
                                       raise AssertionError(self._node_msg(msg))
                                   AssertionError: [node 0] bitcoind should have exited within 2400s with expected error Error: Invalid value detected for '-wallet' or '-nowallet'. '-wallet' requires a string value, while '-nowallet' accepts only '1' to disable all wallets
 test  2024-10-02T00:55:45.690000Z TestFramework (DEBUG): Closing down network thread 
 test  2024-10-02T00:55:45.894000Z TestFramework (INFO): Stopping nodes 
 test  2024-10-02T00:55:45.894000Z TestFramework (WARNING): Not cleaning up dir /Users/runner/work/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20241002_001101/feature_settings_11 
 test  2024-10-02T00:55:45.895000Z TestFramework (ERROR): Test failed. Test logging available at /Users/runner/work/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20241002_001101/feature_settings_11/test_framework.log 
 test  2024-10-02T00:55:45.895000Z TestFramework (ERROR): 
 test  2024-10-02T00:55:45.896000Z TestFramework (ERROR): Hint: Call /Users/runner/work/bitcoin/bitcoin/test/functional/combine_logs.py '/Users/runner/work/bitcoin/bitcoin/ci/scratch/test_runner/test_runner_₿_🏃_20241002_001101/feature_settings_11' to consolidate all logs 
 test  2024-10-02T00:55:45.897000Z TestFramework (ERROR): 
 test  2024-10-02T00:55:45.897000Z TestFramework (ERROR): If this failure happened unexpectedly or intermittently, please file a bug and provide a link or upload of the combined log. 
 test  2024-10-02T00:55:45.897000Z TestFramework (ERROR): https://github.com/bitcoin/bitcoin/issues 
 test  2024-10-02T00:55:45.897000Z TestFramework (ERROR): 
```
</details>

Diff for last comments addressed can be viewed with:
`git range-diff a74bdeea1b8e27b2335f0f7da78006e87ecfb235..f83a4f6929c2287bfd26756a03b096da9c974dfd fc642c33ef28829eda0119a0fe39fd9bc4b84051..b6368fc285bf00b3033061dcd4e29298b227c6df`

luisschwab commented:
Concept ACK

tdb3 commented:
> Please report the CI failure.
Issue #31019 opened

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+
+            if (verbosity <= 0) {
+                for (auto const& orphan : orphanage) {
+                    ret.push_back(orphan.tx->GetHash().ToString()); // txid
```
nit: comment is unnecessary. Also my personal preference is to not put them on the same line

tdb3 replied:
Updated

glozow commented on src/rpc/util.h:
```diff
@@ -100,6 +100,15 @@ uint256 ParseHashO(const UniValue& o, std::string_view strKey);
+ * @param[in] default_verbosity The value to return if verbosity argument is null
+ * @returns An integer describing the verbosity level (e.g. 0, 1, 2, etc.)
+ */
+int ParseVerbosity(const UniValue& arg, int default_verbosity = 0);
```
I find default arguments to be pretty footgunny and prefer to require explicit. Since you're introducing this function for the first time, I suggest no default arg.

```suggestion
int ParseVerbosity(const UniValue& arg, int default_verbosity);
```

hodlinator replied:
Also have a slight preference against default args in new code in general.

tdb3 replied:
Good point. Updated.

glozow commented on src/rpc/blockchain.cpp:
```diff
@@ -766,14 +766,7 @@ static RPCHelpMan getblock()
-            verbosity = request.params[1].getInt<int>();
-        }
-    }
+    int verbosity{ParseVerbosity(request.params[1], 1)};
```
an annotation can help here
```suggestion
   int verbosity{ParseVerbosity(request.params[1], /*default_verbosity=*/1)};
```

tdb3 replied:
Agreed. Updated

glozow commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+
+            UniValue ret(UniValue::VARR);
+
+            if (verbosity <= 0) {
```
It is a bit funky that a negative verbosity turns into 0! I was kind of expecting a `JSONRPCError`. I don't mind and it looks like it's the same for `getblock`, so not asking for a change, but could be worth a test case?

tdb3 replied:
Good point. Test added.

glozow reviewed (commented):
ACK b6368fc285b. I have some picky review comments if you're interested, but I don't feel that strongly on anything except the default argument one.

I tested with @0xB10C's visualizer as well, very cool!
<img width="833" alt="image" src="https://github.com/user-attachments/assets/c1e81c89-3d4d-4a5a-a092-51d633ea10db">

hodlinator reviewed (approved):
ACK b6368fc285bf00b3033061dcd4e29298b227c6df

Reviewed `git range-diff master f83a4f6 b6368fc` + "add tx_in_orphanage" commit + GitHub Files Changed.

Thanks for addressing the latest feedback!

`build/test/functional/rpc_getorphantxs.py` passed locally.
Ran Doxygen and inspected `ParseVerbosity` docs to confirm `[in]`-parameters were handled well etc.

tdb3 commented:
Updated to address comments from @glozow

Removed default argument from `ParseVerbosity()`,  cleaned up comments, and updated the functional test to check negative verbosity.

```
git range-diff fc642c33ef28829eda0119a0fe39fd9bc4b84051..b6368fc285bf00b3033061dcd4e29298b227c6df fc642c33ef28829eda0119a0fe39fd9bc4b84051..98c1536852d1de9a978b11046e7414e79ed40b46
```

tdb3 commented:
> Ran Doxygen and inspected `ParseVerbosity` docs to confirm `[in]`-parameters were handled well etc.
Much appreciated

glozow commented:
reACK 98c1536852d

hodlinator reviewed (approved):
re-ACK 98c1536852d1de9a978b11046e7414e79ed40b46

`git range-diff master b6368fc`:

- Removed default parameter value in `ParseVerbosity` and added `/*default_verbosity=*/` at callsites.
- Added test for `verbosity=-1`.
- Removed "txid"-comment.

danielabrozzoni commented on src/rpc/mempool.cpp:
```diff
@@ -1027,6 +1128,7 @@ void RegisterMempoolRPCCommands(CRPCTable& t)
         {"blockchain", &getrawmempool},
         {"blockchain", &importmempool},
         {"blockchain", &savemempool},
+        {"hidden", &getorphantxs},
```
nit: we could test that it's hidden in the functional test, similarly to `addpeeraddress`/`getrawaddrman` (https://github.com/bitcoin/bitcoin/blob/cfb59da4b3bb34afae467691a3e901f2b5a186f3/test/functional/rpc_net.py#L338-L341)

tdb3 replied:
We seem to be limited/selective in testing for hidden RPCs in functional tests (i.e. there are more hidden RPCs than `addpeeraddress` and `getrawaddrman`).  Maybe that sort of check would/could be a larger scope than just `getorphantxs`?

I'm on the fence about testing that RPCs are hidden.  Changing from hidden to another category would be noticed in a PR review, and having a functional test for it means that more code needs to be maintained.

instagibbs replied:
> Maybe that sort of check would/could be a larger scope than just getorphantxs?

Please don't invalidate the ACKs, but personally find it useful as a reviewer to just enforce expected behavior in tests, even minor ones like this. It's also *documentation* to the future reader as well; the functional tests are often the first place I look for intention of RPCs, without having to dig into git(hub) history to discover it.

danielabrozzoni reviewed (commented):
ACK 98c1536852d1de9a978b11046e7414e79ed40b46

Code looks good to me, I have a few nits, but feel free to ignore them.

pablomartin4btc reviewed (commented):
tACK 98c1536852d1de9a978b11046e7414e79ed40b46

Missed the [club](https://bitcoincore.reviews/30793) but went thru all of it, very interesting and useful. Please set "Review Club" label on the PR so other ppl can go there and learn more about it.

Tested on `mainnet`. Not sure if would be use cases to add other arguments on a follow-up (count, filter by node id), in the meantime I used these queries:
- `./build/src/bitcoin-cli -datadir=${AU_DATADIR} getorphantxs 2 | jq '.[] | select(.from | index(43))`
- `./build/src/bitcoin-cli -datadir=${AU_DATADIR} getorphantxs | jq 'length'
95`
- `./build/src/bitcoin-cli -datadir=${AU_DATADIR} getorphantxs 1 | jq '[.[] | {peer_id: .from[0]}] | group_by(.peer_id) | map({peer_id: .[0].peer_id, count: length})'`
  ```
  [
    {
      "peer_id": 43,
      "count": 95
    }
  ]
  ```

The code looks very clean and neat.

Also tested the [treemap](https://observablehq.com/d/a481f4ced64b7975) made by @0xB10C.

tdb3 commented on test/functional/rpc_getorphantxs.py:
```diff
@@ -0,0 +1,130 @@
+        assert_equal(node.getmempoolinfo()["size"], 0)
+
+        self.log.info("Check that both children are in the orphanage")
+
```
True.  I'll leave this for a potential follow-up

tdb3 commented:
Thanks for the review @danielabrozzoni and @pablomartin4btc!

> Tested on `mainnet`. Not sure if would be use cases to add other arguments on a follow-up (count, filter by node id)

Yes, those sound interesting for a follow-up or something like `getorphanageinfo`.

itornaza reviewed (approved):
reACK 98c1536852d1de9a978b11046e7414e79ed40b46

davidgumberg commented on src/rpc/mempool.cpp:
```diff
@@ -812,6 +815,104 @@ static RPCHelpMan savemempool()
+        "\nEXPERIMENTAL warning: this call may be changed in future releases.\n",
+        {
+            {"verbosity|verbose", RPCArg::Type::NUM, RPCArg::Default{0}, "0 for an array of txids (may contain duplicates), 1 for an array of objects with tx details, and 2 for details from (1) and tx hex",
+             RPCArgOptions{.skip_type_check = true}},
```
non-blocking out-of-scope: Might be worth it to either deprecate true/false values for verbosity arguments or introduce an argument type specific to verbosity arguments to avoid skipping type checks for them.

tdb3 replied:
I would support either, with the latter being my preference (at least initially/in transition)

rkrux reviewed (commented):
Post Merge tACK 98c1536852d1de9a978b11046e7414e79ed40b46

Successful make and functional tests.

I tried to test this RPC in my local by connecting 2 nodes in the regtest env. However, I couldn't get an orphan transaction generated. I tried 2 ways:
1. Set `mempoolexpiry` to 1hr in one node and broadcast the child transaction 15mins after the parent transaction from the other node. Initially I expected the parent transaction to get dropped after an hour but now it didn't. My guess it's because of the presence of atleast 1 descendant.
2. Propagated both the transactions from the other node and deleted the `mempool.dat` file of the first node. Restarted the first node and broadcast the child transaction from the second node but the parent transaction was also implicitly broadcast to the first node.

Is there any other easy way to generate orphans?
